tests/ui/test_fastapi.py

```diff
@@ -0,0 +1,246 @@
+        rewards_prompt="Rewards Prompt",
+        tag="test_tag",
+    )
+    episode_log.save()
+    pk = episode_log.pk
+    assert pk is not None
+    return pk
+
+
+def delete_mock_episode_log(pk: str) -> None:
+    EpisodeLog.delete(pk)
+
+
+def test_get_scenarios_all() -> None:
+    pk = create_mock_env_profile()
+    response = client.get("/scenarios")
+    delete_mock_env_profile(pk)
+    assert response.status_code == 200
+    assert isinstance(response.json(), list)
+
+
+def test_get_scenarios_by_id() -> None:
+    pk = create_mock_env_profile()
+    response = client.get(f"/scenarios/id/{pk}")
+    delete_mock_env_profile(pk)
+    assert response.status_code == 200
+    assert isinstance(response.json(), list)
+    assert response.json()[0]["scenario"] == "A"
+
+
+def test_get_scenarios_by_codename() -> None:
+    codename = "test_codename"
+    pk = create_mock_env_profile()
+    response = client.get(f"/scenarios/codename/{codename}")
+    delete_mock_env_profile(pk)
+    assert response.status_code == 200
+    assert isinstance(response.json(), list)
+    assert response.json()[0]["codename"] == codename
+
+
+def test_get_agents_all() -> None:
+    pk = create_mock_agent_profile()
+    response = client.get("/agents")
+    delete_mock_agent_profile(pk)
+    assert response.status_code == 200
+    assert isinstance(response.json(), list)
+
+
+def test_get_agents_by_id() -> None:
+    pk = create_mock_agent_profile()
+    response = client.get(f"/agents/id/{pk}")
+    delete_mock_agent_profile(pk)
+    assert response.status_code == 200
+    assert isinstance(response.json(), list)
+    assert response.json()[0]["first_name"] == "John"
+
+
+def test_get_agents_by_gender() -> None:
+    gender = "test_gender"
+    pk = create_mock_agent_profile()
+    response = client.get(f"/agents/gender/{gender}")
+    delete_mock_agent_profile(pk)
+    assert response.status_code == 200
+    assert isinstance(response.json(), list)
+    assert response.json()[0]["gender"] == gender
+
+
+def test_get_agents_by_occupation() -> None:
+    occupation = "test_occupation"
+    pk = create_mock_agent_profile()
+    response = client.get(f"/agents/occupation/{occupation}")
+    delete_mock_agent_profile(pk)
+    assert response.status_code == 200
+    assert isinstance(response.json(), list)
+    assert response.json()[0]["occupation"] == occupation
+
+
+def test_get_episodes_by_id() -> None:
+    pk = create_mock_episode_log()
+    response = client.get(f"/episodes/id/{pk}")
+    delete_mock_episode_log(pk)
+    assert response.status_code == 200
+    assert isinstance(response.json(), list)
+
+
+def test_get_episodes_by_tag() -> None:
+    tag = "test_tag"
+    pk = create_mock_episode_log()
+    response = client.get(f"/episodes/tag/{tag}")
+    delete_mock_episode_log(pk)
+    assert response.status_code == 200
+    assert isinstance(response.json(), list)
+    assert response.json()[0]["tag"] == tag
+
+
+def test_create_agent() -> None:
```

Comment:
@ProKil This thing is hard to make it work, any chance you may take a look?

Reply:
`uv run pytest -s tests/ui/test_fastapi.py::test_create_agent`

error msg
```
================================================= FAILURES ==================================================
_____________________________________________ test_create_agent _____________________________________________

self = <pydantic.type_adapter.TypeAdapter object at 0x10872fa90>, rebuild_mocks = False

    @_frame_depth(1)
    def _init_core_attrs(self, rebuild_mocks: bool) -> None:
        try:
>           self._core_schema = _getattr_no_parents(self._type, '__pydantic_core_schema__')

.venv/lib/python3.11/site-packages/pydantic/type_adapter.py:270: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

obj = typing.Annotated[typing.Union[str, NoneType, redis_om.model.model.ExpressionProxy], FieldInfo(annotation=Union[str, No...(annotation=Union[str, NoneType], required=False, default=None, alias='pk', validate_default=True, metadata=[...])])])]
attribute = '__pydantic_core_schema__'

    def _getattr_no_parents(obj: Any, attribute: str) -> Any:
        """Returns the attribute value without attempting to look up attributes from parent types."""
        if hasattr(obj, '__dict__'):
            try:
                return obj.__dict__[attribute]
            except KeyError:
                pass
    
        slots = getattr(obj, '__slots__', None)
        if slots is not None and attribute in slots:
            return getattr(obj, attribute)
        else:
>           raise AttributeError(attribute)
E           AttributeError: __pydantic_core_schema__

.venv/lib/python3.11/site-packages/pydantic/type_adapter.py:112: AttributeError

During handling of the above exception, another exception occurred:

    def test_create_agent() -> None:
        agent_data = {"first_name": "test_first_name", "last_name": "test_last_name", "tag": "test"}
        #agent_data = {"name": "test_name", "description": "test_description", "price": 100, "tax": 10}
>       response = client.post("/agents/", json=agent_data)

tests/ui/test_fastapi.py:180: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.11/site-packages/starlette/testclient.py:597: in post
    return super().post(
.venv/lib/python3.11/site-packages/httpx/_client.py:1157: in post
    return self.request(
.venv/lib/python3.11/site-packages/starlette/testclient.py:488: in request
    return super().request(
.venv/lib/python3.11/site-packages/httpx/_client.py:837: in request
    return self.send(request, auth=auth, follow_redirects=follow_redirects)
.venv/lib/python3.11/site-packages/httpx/_client.py:926: in send
    response = self._send_handling_auth(
.venv/lib/python3.11/site-packages/httpx/_client.py:954: in _send_handling_auth
    response = self._send_handling_redirects(
.venv/lib/python3.11/site-packages/httpx/_client.py:991: in _send_handling_redirects
    response = self._send_single_request(request)
.venv/lib/python3.11/site-packages/httpx/_client.py:1027: in _send_single_request
    response = transport.handle_request(request)
.venv/lib/python3.11/site-packages/starlette/testclient.py:381: in handle_request
    raise exc
.venv/lib/python3.11/site-packages/starlette/testclient.py:378: in handle_request
    portal.call(self.app, scope, receive, send)
.venv/lib/python3.11/site-packages/anyio/from_thread.py:290: in call
    return cast(T_Retval, self.start_task_soon(func, *args).result())
../../miniconda3/envs/sotopia/lib/python3.11/concurrent/futures/_base.py:449: in result
    return self.__get_result()
../../miniconda3/envs/sotopia/lib/python3.11/concurrent/futures/_base.py:401: in __get_result
    raise self._exception
.venv/lib/python3.11/site-packages/anyio/from_thread.py:221: in _call_func
    retval = await retval_or_awaitable
.venv/lib/python3.11/site-packages/fastapi/applications.py:1054: in __call__
    await super().__call__(scope, receive, send)
.venv/lib/python3.11/site-packages/starlette/applications.py:113: in __call__
    await self.middleware_stack(scope, receive, send)
.venv/lib/python3.11/site-packages/starlette/middleware/errors.py:187: in __call__
    raise exc
.venv/lib/python3.11/site-packages/starlette/middleware/errors.py:165: in __call__
    await self.app(scope, receive, _send)
.venv/lib/python3.11/site-packages/starlette/middleware/exceptions.py:62: in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
.venv/lib/python3.11/site-packages/starlette/_exception_handler.py:62: in wrapped_app
    raise exc
.venv/lib/python3.11/site-packages/starlette/_exception_handler.py:51: in wrapped_app
    await app(scope, receive, sender)
.venv/lib/python3.11/site-packages/starlette/routing.py:715: in __call__
    await self.middleware_stack(scope, receive, send)
.venv/lib/python3.11/site-packages/starlette/routing.py:735: in app
    await route.handle(scope, receive, send)
.venv/lib/python3.11/site-packages/starlette/routing.py:288: in handle
    await self.app(scope, receive, send)
.venv/lib/python3.11/site-packages/starlette/routing.py:76: in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
.venv/lib/python3.11/site-packages/starlette/_exception_handler.py:62: in wrapped_app
    raise exc
.venv/lib/python3.11/site-packages/starlette/_exception_handler.py:51: in wrapped_app
    await app(scope, receive, sender)
.venv/lib/python3.11/site-packages/starlette/routing.py:73: in app
    response = await f(request)
.venv/lib/python3.11/site-packages/fastapi/routing.py:291: in app
    solved_result = await solve_dependencies(
.venv/lib/python3.11/site-packages/fastapi/dependencies/utils.py:656: in solve_dependencies
    ) = await request_body_to_args(  # body_params checked above
.venv/lib/python3.11/site-packages/fastapi/dependencies/utils.py:881: in request_body_to_args
    fields_to_extract = get_cached_model_fields(first_field.type_)
.venv/lib/python3.11/site-packages/fastapi/_compat.py:657: in get_cached_model_fields
    return get_model_fields(model)
.venv/lib/python3.11/site-packages/fastapi/_compat.py:284: in get_model_fields
    return [
.venv/lib/python3.11/site-packages/fastapi/_compat.py:285: in <listcomp>
    ModelField(field_info=field_info, name=name)
<string>:6: in __init__
    ???
.venv/lib/python3.11/site-packages/fastapi/_compat.py:110: in __post_init__
    self._type_adapter: TypeAdapter[Any] = TypeAdapter(
.venv/lib/python3.11/site-packages/pydantic/type_adapter.py:257: in __init__
    self._init_core_attrs(rebuild_mocks=False)
.venv/lib/python3.11/site-packages/pydantic/type_adapter.py:135: in wrapped
    return func(self, *args, **kwargs)
.venv/lib/python3.11/site-packages/pydantic/type_adapter.py:277: in _init_core_attrs
    self._core_schema = _get_schema(self._type, config_wrapper, parent_depth=self._parent_depth)
.venv/lib/python3.11/site-packages/pydantic/type_adapter.py:95: in _get_schema
    schema = gen.generate_schema(type_)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:655: in generate_schema
    schema = self._generate_schema_inner(obj)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:908: in _generate_schema_inner
    return self._annotated_schema(obj)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2028: in _annotated_schema
    schema = self._apply_annotations(source_type, annotations)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2107: in _apply_annotations
    schema = get_inner_schema(source_type)
.venv/lib/python3.11/site-packages/pydantic/_internal/_schema_generation_shared.py:83: in __call__
    schema = self._handler(source_type)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2189: in new_handler
    schema = metadata_get_schema(source, get_inner_schema)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2185: in <lambda>
    lambda source, handler: handler(source)
.venv/lib/python3.11/site-packages/pydantic/_internal/_schema_generation_shared.py:83: in __call__
    schema = self._handler(source_type)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2189: in new_handler
    schema = metadata_get_schema(source, get_inner_schema)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2185: in <lambda>
    lambda source, handler: handler(source)
.venv/lib/python3.11/site-packages/pydantic/_internal/_schema_generation_shared.py:83: in __call__
    schema = self._handler(source_type)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2088: in inner_handler
    schema = self._generate_schema_inner(obj)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:929: in _generate_schema_inner
    return self.match_type(obj)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1029: in match_type
    return self._match_generic_type(obj, origin)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1058: in _match_generic_type
    return self._union_schema(obj)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1378: in _union_schema
    choices.append(self.generate_schema(arg))
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:655: in generate_schema
    schema = self._generate_schema_inner(obj)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:929: in _generate_schema_inner
    return self.match_type(obj)
.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1038: in match_type
    return self._unknown_type_schema(obj)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pydantic._internal._generate_schema.GenerateSchema object at 0x108731580>
obj = <class 'redis_om.model.model.ExpressionProxy'>

    def _unknown_type_schema(self, obj: Any) -> CoreSchema:
>       raise PydanticSchemaGenerationError(
            f'Unable to generate pydantic-core schema for {obj!r}. '
            'Set `arbitrary_types_allowed=True` in the model_config to ignore this error'
            ' or implement `__get_pydantic_core_schema__` on your type to fully support it.'
            '\n\nIf you got this error by calling handler(<some type>) within'
            ' `__get_pydantic_core_schema__` then you likely need to call'
            ' `handler.generate_schema(<some type>)` since we do not call'
            ' `__get_pydantic_core_schema__` on `<some type>` otherwise to avoid infinite recursion.'
        )
E       pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class 'redis_om.model.model.ExpressionProxy'>. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.
E       
E       If you got this error by calling handler(<some type>) within `__get_pydantic_core_schema__` then you likely need to call `handler.generate_schema(<some type>)` since we do not call `__get_pydantic_core_schema__` on `<some type>` otherwise to avoid infinite recursion.
E       
E       For further information visit https://errors.pydantic.dev/2.9/u/schema-for-unknown-type

.venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:558: PydanticSchemaGenerationError
============================================= warnings summary ==============================================
.venv/lib/python3.11/site-packages/starlette/formparsers.py:12
  /Users/xuhuizhou/Projects/sotopia/.venv/lib/python3.11/site-packages/starlette/formparsers.py:12: PendingDeprecationWarning: Please use `import python_multipart` instead.
    import multipart

.venv/lib/python3.11/site-packages/redis_om/model/model.py:1502
  /Users/xuhuizhou/Projects/sotopia/.venv/lib/python3.11/site-packages/redis_om/model/model.py:1502: PydanticDeprecatedSince20: Pydantic V1 style `@validator` validators are deprecated. You should migrate to Pydantic V2 style `@field_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.9/migration/
    @validator("pk", always=True, allow_reuse=True)

.venv/lib/python3.11/site-packages/pydantic/deprecated/class_validators.py:121
  /Users/xuhuizhou/Projects/sotopia/.venv/lib/python3.11/site-packages/pydantic/deprecated/class_validators.py:121: PydanticDeprecatedSince20: `allow_reuse` is deprecated and will be ignored; it should no longer be necessary. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.9/migration/
    warn(_ALLOW_REUSE_WARNING_MESSAGE, DeprecationWarning)

.venv/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py:293: 42 warnings
  /Users/xuhuizhou/Projects/sotopia/.venv/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py:293: PydanticDeprecatedSince20: The `__fields__` attribute is deprecated, use `model_fields` instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.9/migration/
    warnings.warn(

.venv/lib/python3.11/site-packages/pydantic/_internal/_fields.py:132
  /Users/xuhuizhou/Projects/sotopia/.venv/lib/python3.11/site-packages/pydantic/_internal/_fields.py:132: UserWarning: Field "model_id" in AgentProfileWithPersonalInformation has conflict with protected namespace "model_".
  
  You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
```

Reply:
I suspect it's due to the jsonmodel thing and pydantic V2.

below stuff works, but when I switch to AgentProfile, things go wrong.
```
from fastapi import FastAPI
from pydantic import BaseModel


class Item(BaseModel):
    name: str
    description: str | None = None
    price: float
    tax: float | None = None


app = FastAPI()


@app.post("/items/")
async def create_item(item: Item):
    return item
```

Reply:
As mentioned in the error message: Set `arbitrary_types_allowed=True` in the model_config to ignore this error

Reply:
The error log in CI is a different error b/c the PUT endpoint is not implemented.

Reply:
See https://github.com/sotopia-lab/sotopia/actions/runs/11963283235/job/33353427057?pr=246#step:4:230